File: app/javascript/channels/chat_room_channel.js
import consumer from "./consumer"
const path = window.location.pathname;
const ROOM_ID = path.split("/").pop();

const chatRoom = consumer.subscriptions.create({channel: "ChatRoomChannel", chat_room_id: ROOM_ID, user_id: currentUserId}, {
  connected() {
    // Called when the subscription is ready for use on the server
  },

  disconnected() {
    // Called when the subscription has been terminated by the server
  },

  received(data) {
    const messages = document.getElementById('messages');
    let messageHTML
    if(data.user_id.toString() == currentUserId){
      messageHTML = data.message;
    }else{
      messageHTML = data.second_message;
    }
    messages.insertAdjacentHTML('beforeend', messageHTML);
  },

  speak: function(message) {
    return this.perform('speak', { message: message });
  }
});

window.addEventListener("keypress", function(event) {
  if (event.key === 'Enter') {
    chatRoom.speak(event.target.value);
    event.target.value = '';
    event.preventDefault();
  }
});

File: app/channels/chat_room_channel.rb
class ChatRoomChannel < ApplicationCable::Channel
  def subscribed
    stream_from "chat_room_channel_#{params[:chat_room_id]}"
  end

  def unsubscribed
    # Any cleanup needed when channel is unsubscribed
  end

  def speak(data)
    message = ChatMessage.create! message: data['message'], user_id: current_user.id, chat_room_id: params[:chat_room_id], send_at: Time.now
    ChatMessageBroadcastJob.perform_later(message)
  end
end
